homeassistant/components/glinet/router.py

```diff
@@ -0,0 +1,325 @@
+    CONF_CONSIDER_HOME,
+    DEFAULT_CONSIDER_HOME,
+    DOMAIN as TRACKER_DOMAIN,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_API_TOKEN, CONF_HOST, CONF_PASSWORD
+from homeassistant.core import (  # callback,CALLBACK_TYPE
+    CALLBACK_TYPE,
+    HomeAssistant,
+    callback,
+)
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_registry import RegistryEntry
+from homeassistant.helpers.event import async_track_time_interval
+
+# from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util import dt as dt_util
+
+from .const import DOMAIN
+
+# from typing import Any
+
+
+# from homeassistant.helpers.event import async_track_time_interval
+
+
+_LOGGER = logging.getLogger(__name__)
+SCAN_INTERVAL = timedelta(seconds=30)
+
+
+class ClientDevInfo:
+    """Representation of a device connected to the router."""
+
+    def __init__(self, mac: str, name=None):
+        """Initialize a connected device."""
+        self._mac: str = mac
+        self._name: str | None = name
+        self._ip_address: str | None = None
+        self._last_activity: datetime = dt_util.utcnow() - timedelta(days=1)
+        self._connected: bool = False
+
+    def update(self, dev_info: dict = None, consider_home=0):
+        """Update connected device info."""
+        # TODO actually cause this function to run by configuring some devices to track
+        now: datetime = dt_util.utcnow()
+        if dev_info:
+            if not self._name:
+                pass  # TODO access device name - GLinet usus '*' if unknown
+                if dev_info["name"] == "*":
+                    self._name = self._mac.replace(":", "_")
+                else:
+                    self._name = dev_info["name"]
+            self._ip_address = dev_info["ip"]
+            self._last_activity = now
+            self._connected = dev_info["online"]
+
+        # a device might not actually be online but we want to consider it home
+        elif self._connected:
+            self._connected = (
+                now - self._last_activity
+            ).total_seconds() < consider_home
+            self._ip_address = None
+
+    @property
+    def is_connected(self):
+        """Return connected status."""
+        return self._connected
+
+    @property
+    def mac(self):
+        """Return device mac address."""
+        return self._mac
+
+    @property
+    def name(self):
+        """Return device name."""
+        return self._name
+
+    @property
+    def ip_address(self):
+        """Return device ip address."""
+        return self._ip_address
+
+    @property
+    def last_activity(self):
+        """Return device last activity."""
+        return self._last_activity
+
+
+class GLinetRouter:
```

Comment:
Isn't it better to move some api related stuff to your Glinet library? I feel the code and communication is a bit all over the place?

Reply:
I see what you mean. I was trying to keep the python library as lean as possible, it is just a wrapper for async requests really, I could move some of the functions and properties over but my worry is that the codebase would become rather 'split' and would move so code out of the reach of the home assistant maintainers and admins.

Reply:
As you may have picked up, much of the code is borrowed from the ASUS router integration and so I have tried to replicate that as much as I can - I am happy to try and improve the 'smell' of the code with the reviewers' guidance